theano/gof/destroyhandler.py

```diff
@@ -777,20 +781,68 @@ def on_detach(self, fgraph):
         delattr(self.fgraph, 'destroy_handler')
         self.fgraph = None
 
+    def fast_destroy(self, app, reason):
+        """
+        Do the check for only 1 level.
+
+        For now:
+        - Destroyed variables can have only 1 clients.
+        - Allow sequence of destroy variables.
+        - Allow view to have multiple clients.
+        - Allow sequence of view.
+        - But don't allow to destroy view
+        """
+        dm = getattr(app.op, 'destroy_map', None)
+        if not dm:
+            return
+        inputs = set(itertools.chain.from_iterable(dm.values()))   # list of app's destroyed inputs
+        for inp_idx in inputs:
+            inp = app.inputs[inp_idx]
+            if getattr(inp.tag, 'indestructible', False) or isinstance(inp, graph.Constant):
+                self.fail_validate[app] = InconsistencyError(
+                    "Attempting to destroy indestructible variables: %s" %
+                    inp)
+            elif len(inp.clients) > 1:
+                self.fail_validate[app] = theano.gof.InconsistencyError(
+                    "Destroyed variable has more than one client. " + str(reason))
+            elif inp.owner:
+                app2 = inp.owner
+                inp_idx2 = app2.outputs.index(inp)
+                v = getattr(app2.op, 'view_map', {})
+                d = getattr(app2.op, 'destroy_map', {})
+                if v:
+                    v = v.get(inp_idx2, [])
+                    if len(v) > 0:
+                        self.fail_validate[app] = theano.gof.InconsistencyError(
+                            "Destroyed variable has view_map. " + str(reason))
+                    elif d:
+                        d = d.get(inp_idx2, [])
+                        if len(d) > 0:
+                            self.fail_validate[app] = theano.gof.InconsistencyError(
+                                "Destroyed variable has destroy_map. " + str(reason))
+
+                # These 2 assertions are commented since this function is called so many times
+                # but they should be true.
+                # assert len(v) <= 1
+                # assert len(d) <= 1
+
     def on_import(self, fgraph, app, reason):
         """
         Add Apply instance to set which must be computed.
 
         """
-
         if app in self.debug_all_apps:
             raise ProtocolError("double import")
         self.debug_all_apps.add(app)
         # print 'DH IMPORT', app, id(app), id(self), len(self.debug_all_apps)
 
         # If it's a destructive op, add it to our watch list
-        if getattr(app.op, 'destroy_map', {}):
+        if getattr(app.op, 'destroy_map', None):
             self.destroyers.add(app)
+            if self.algo == 'fast':
+                self.fast_destroy(app, reason)
+        elif getattr(app.op, 'view_map', None) and self.algo == 'fast':
+            self.fast_destroy(app, reason)
```

Comment:
@ReyhaneAskari if you remember why this is here, add a comment. I forgot why it is needed.